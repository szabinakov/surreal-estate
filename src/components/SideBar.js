import React, { useState } from 'react'
import {Link, useLocation, useHistory} from 'react-router-dom'
import '../styles/SideBar.css'
import qs from 'qs'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {faSearch} from '@fortawesome/free-solid-svg-icons'


const SideBar = () => {
    const [query, setQuery ] = useState('')
    const { search } = useLocation()
    const history = useHistory()
    
    const buildQueryString = (operation, valueObj) => {
        
        const currentQueryParams = qs.parse(search, { ignoreQueryPrefix : true } )
        
        const newQueryParams = {
            ...currentQueryParams,
             [operation]: JSON.stringify({
                 ...JSON.parse(currentQueryParams[operation] || '{}'),
                 ...valueObj,
             })
        }
        return qs.stringify(newQueryParams, { addQueryPrefix: true, encode : false })
    }

    const handleSearch = (event) => {
        event.preventDefault()
        const newQuery = buildQueryString('query', { title: { $regex: query } })
    
        history.push(newQuery)
    }

    return (

        <div className='SideBar'>
            <form onSubmit={handleSearch} className='SearchForm'>
                <input 
                    value={query}
                    onChange={(event) => setQuery(event.target.value)}
                    type='text' 
                    placeholder='Search!' 
                    className='SearchInput'/>

                    <button
                        type='submit' 
                        className='SearchButton'>
                            Search <FontAwesomeIcon icon={faSearch}/>
                    </button>
            </form>
            <h3>Filter By City</h3>
            <div className='FilterByCity'>
            <p>
                <Link 
                to={buildQueryString('query', {city: 'Manchester'})}
                className='SearchCityLink'>
                    Manchester</Link>
            </p>
            <p>
                <Link 
                to={buildQueryString('query', {city: 'Leeds'})}
                className='SearchCityLink'>
                    Leeds</Link>
            </p>
            <p>
                <Link 
                to={buildQueryString('query', {city: 'Sheffield'})}
                className='SearchCityLink'
                >Sheffield</Link>
            </p>
            <p>
                <Link 
                to={buildQueryString('query', {city: 'Liverpool'})}
                className='SearchCityLink'>
                    Liverpool</Link>
            </p>
            </div>
            
                <h3>Sort By</h3>
            <div className='SortBy'>
            <p>
                <Link 
                to={buildQueryString('sort',  { price: 1 })}
                className='SortPriceLink'>
                    Price Ascending</Link>
            </p>
            <p>
                <Link 
                to={buildQueryString('sort', { price: -1 })}
                className='SortPriceLink'>
                    Price Descending</Link>
            </p>
            </div>
        </div>

    )
   
}


export default SideBar