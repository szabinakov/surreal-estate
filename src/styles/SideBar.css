.SideBar {

    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
}
p {
    padding: 10px;
}