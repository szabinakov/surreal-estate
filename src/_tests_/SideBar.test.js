import React from 'react'
import { render } from '@testing-library/react';
import {MemoryRouter} from 'react-router-dom'
import SideBar from '../components/SideBar';

describe('SideBar', () => {

    it('matches the snapshot', () => {
        const {asFragment} = render (
        <MemoryRouter>
        <SideBar/>
        </MemoryRouter>)

        expect(asFragment()).toMatchSnapshot()
    })
})