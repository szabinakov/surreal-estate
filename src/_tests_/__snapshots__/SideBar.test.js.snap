// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="SideBar"
  >
    <form
      class="SearchForm"
    >
      <input
        class="SearchInput"
        placeholder="Search!"
        type="text"
        value=""
      />
      <button
        class="SearchButton"
        type="submit"
      >
        Search 
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
          />
        </svg>
      </button>
    </form>
    <h3>
      Filter By City
    </h3>
    <div
      class="FilterByCity"
    >
      <a
        class="SearchCityLink"
        href="/?query={\\"city\\":\\"Manchester\\"}"
      >
        Manchester
      </a>
      <a
        class="SearchCityLink"
        href="/?query={\\"city\\":\\"Leeds\\"}"
      >
        Leeds
      </a>
      <a
        class="SearchCityLink"
        href="/?query={\\"city\\":\\"Sheffield\\"}"
      >
        Sheffield
      </a>
      <a
        class="SearchCityLink"
        href="/?query={\\"city\\":\\"Liverpool\\"}"
      >
        Liverpool
      </a>
    </div>
    <h3>
      Sort By
    </h3>
    <div
      class="SortBy"
    >
      <a
        class="SortPriceLink"
        href="/?sort={\\"price\\":1}"
      >
        Price Ascending
      </a>
      <a
        class="SortPriceLink"
        href="/?sort={\\"price\\":-1}"
      >
        Price Descending
      </a>
    </div>
  </div>
</DocumentFragment>
`;
